### ISSUE TITLE
Document backend_config.ini

### ISSUE THREAD

**Issue by matsduf** [zonemaster/zonemaster-backend#623]
All settings in backend_config.ini are not documented in [Configuration.md], e.g. ZONEMASTER.force_hash_id_use_in_API_starting_from_id is not.

[Configuration.md]: https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md

**Comment by pnax** [comment]
There are 2 last keys to document :
* `number_of_processes_for_frontend_testing`
* `number_of_processes_for_batch_testing`
The names used for the keys seem to express the fact that we can use these keys to limit the number of processes depending on the usage.

However looking at the current code, I can't find much on this potential limitation. What I found is the following usage, where both values are used simultaneously to compute the maximum number of processes.

https://github.com/zonemaster/zonemaster-backend/blob/5b485f7f88caa545150e011fdb835b7767941b30/lib/Zonemaster/Backend/Config.pm#L359..L410

I then tried to look for code that can limit the creation of a process for each of the usage (frontend or batch) without success.

Going through the git history I was able to find code where the limitation was explicit (this is quite old, but this is around the time where the `NumberOfProfessesForFrontendTesting` and `NumberOfProfessesForBatchTesting` (sic) were added - https://github.com/zonemaster/zonemaster-backend/commit/4df3d9bd8ac43e8ddfccf4df7aca9b7ed2acc699)

https://github.com/zonemaster/zonemaster-backend/blob/4df3d9bd8ac43e8ddfccf4df7aca9b7ed2acc699/JobRunner/execute_tests.pl#L68..L84

https://github.com/zonemaster/zonemaster-backend/blob/4df3d9bd8ac43e8ddfccf4df7aca9b7ed2acc699/JobRunner/execute_tests.pl#L93

So it looks like the file `execute_tests.pl` was using the limitation per type of process. This file is no more in the codebase. And it seems that its substitute has not kept the logic on frontend and batch processes (see commit https://github.com/zonemaster/zonemaster-backend/commit/55f72d9ead3ab0e06f81b4a9d390d0972ebe14de)

https://github.com/zonemaster/zonemaster-backend/blob/55f72d9ead3ab0e06f81b4a9d390d0972ebe14de/script/zm_wb_daemon#L42..L44


Unless I am missing something it looks like the idea to limit the number of processes per usage has not been kept into the code. So I suggest we decide what we do about batch processing as proposed in #743.

The best I come with at this point is to specify that these values are added together and the result defines the total number of allowed processes. Despite there names, they do not limit the number of process by type (frontend or batch).

**Comment by pnax** [comment]
@matsduf and @mattias-p can you have a look at my above comment ? (I don't know if editing the comment with @ mentions raises a notification on your sides)

**Comment by matsduf** [comment]
I think it would be fine to deprecate both `number_of_processes_for_frontend_testing` and `number_of_processes_for_batch_testing`, and replace that with a new key with a shorter name for limiting the number of processes. We should also set a reasonable default value on that so that the default config.ini can be without the key. I guess 10 could be such a value.

I think it is unclear, however, what will happen when you reach the limit.

Before we do anything about #743 we should contact RIPE that we could be using the batch function.

**Comment by mattias-p** [comment]
> I think it would be fine to deprecate both number_of_processes_for_frontend_testing and number_of_processes_for_batch_testing, and replace that with a new key with a shorter name for limiting the number of processes.

I agree with this. If you want two test agent worker pools it seems more reasonable to just run two test agent daemons.

> We should also set a reasonable default value on that so that the default config.ini can be without the key. I guess 10 could be such a value.

I tried to look around to see what other people use. In one page of search engine hits I found two examples. The first one I found, [Gunicorn](https://docs.gunicorn.org/en/stable/settings.html#workers), defaults to 1. The second one, [parpool](https://se.mathworks.com/help/parallel-computing/parpool.html) (some MatLab parallelization library or I don't know), defaults to one worker per physical core.

Having a default value of 1 worker for the test agent makes sense to me. One worker per core would make sense to me in a way if the test utilization is close to 100%. From looking at `top` during a single test I got a CPU utilization of around 50%. Another thing to consider is that if the test agent workers are hogging too much CPU the rpcapi workers would also be affected.

> I think it is unclear, however, what will happen when you reach the limit.

Unclear in what way? Do you mean that there might be bugs in Parallel::ForkManager?

**Comment by matsduf** [comment]
> > I think it is unclear, however, what will happen when you reach the limit.
> 
> Unclear in what way? Do you mean that there might be bugs in Parallel::ForkManager?

Well, unclear to me, I guess. :-)

**Comment by mattias-p** [comment]
The way I read the design is that Parallel::ForkManager grows its worker pool until it reaches its given limit. If new test requests come in at a higher rate than the worker pool can handle them, the test requests start queuing up. The queue is handled in FIFO order.

**Comment by matsduf** [comment]
@mattias-p, thank you!

**Comment by pnax** [comment]
I think that all settings are now properly documented. Can we close this issue?

**Comment by matsduf** [comment]
Fine to close. If we find something new that is properly documented it is better to open a new issue and specify what is missing.
